### PR TITLE
Implement IonSerializers option for custom serializers

### PR DESCRIPTION
**Pull request by justing-bq**
Resolves https://github.com/amzn/ion-object-mapper-dotnet/issues/14.

IonSerializers: `Dictionary<Type, IonSerializer>`

A Dictionary of IonSerializers which specifies, for a key Type, a custom Ion serializer for that type.

We will provide extension points to allow for arbitrarily fine-grained control over the serialization process. There are a few ways to specify this. To the IonSerializer, one can pass in options including a Dictionary of Type onto IonSerializer. This mapping will take precedence over the default mapping.

```c#
new IonSerializer(new IonSerializationOptions
{
     IonSerializers = new Dictionary<Type, IonSerializer>()
    {
        {typeof(string), new MyCustomStringIonSerializer()},
        {typeof(DateTime), new MyCustomDateTimeIonSerializer()}
    },
});
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 1]
```diff
@@ -286,89 +243,133 @@ public object Deserialize(IIonReader reader, Type type)
...
 
-            throw new NotSupportedException("Don't know how to Deserialize this Ion data. Last IonType was: " + ionType);
+            return false;
         }
 
-        public T Deserialize<T>(IIonReader reader)
+        private void SerializePrimitive(Type type, IIonWriter writer, object item)
         {
-            return (T) Deserialize(reader, typeof(T));
+            if (type == typeof(bool)) this.primitiveSerializers[type].Serialize(writer, Convert.ToBoolean(item));
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:350]: Annoyingly you can't switch on `typeof()` for primitive types.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 2]
```diff
@@ -83,6 +84,52 @@ public void DeserializesObjectsWithIgnoreDefaults()
...
+            Assert.IsTrue(serializedCar.ContainsField("make"));
+            Assert.AreEqual(TestObjects.honda.Make.ToUpper(), serializedCar.GetField("make").StringValue);
+            Assert.IsTrue(serializedCar.ContainsField("yearOfManufacture"));
+            Assert.AreEqual(-TestObjects.honda.YearOfManufacture, serializedCar.GetField("yearOfManufacture").IntValue);
+            
+            stream = defaultSerializer.Serialize(TestObjects.honda);
+            var deserializedCar = customSerializer.Deserialize<Car>(stream);
+            Assert.AreEqual(TestObjects.honda.Make.ToUpper(), deserializedCar.Make);
+            Assert.AreEqual(-TestObjects.honda.YearOfManufacture, deserializedCar.YearOfManufacture);
+        }
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:133]: Please add tests for failure cases (e.g. when the custom serializer fails the validation.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 3]
```diff
@@ -163,117 +205,34 @@ public void Serialize<T>(IIonWriter writer, T item)
...
+            Type type = item.GetType();
+            if (this.primitiveSerializers.ContainsKey(type))
             {
-                new IonBooleanSerializer().Serialize(writer, Convert.ToBoolean(item));
+                this.SerializePrimitive(type, writer, item);
                 return;
             }
 
-            if (item is int)
+            if (item is IList)
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:219]: New C# compiler supports type in switch, can we use it here.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch

File: Amazon.Ion.ObjectMapper.Test/TestObjects.cs [hunk 4]
```diff
@@ -286,4 +287,30 @@ public override string ToString()
...
+        {
+            return -reader.IntValue();
+        }
+
+        public void Serialize(IIonWriter writer, int item)
+        {
+            writer.WriteInt(-item);
+        }
+    }
 }
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper.Test/TestObjects.cs:316]: Please add tests for custom serializers for other types.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 5]
```diff
@@ -288,87 +247,176 @@ public object Deserialize(IIonReader reader, Type type, IonType ionType)
...
+
+            else if (type == typeof(SymbolToken))
+                serializer.Serialize(writer, (SymbolToken)item);
+            
+            else if (type == typeof(DateTime))
+                serializer.Serialize(writer, (DateTime)item);
+
+            else if (type == typeof(Guid))
+                serializer.Serialize(writer, (Guid)item);
+        }
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:420]: New C# compiler supports type in switch, can we use it here.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch

**Review by HomayounA — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 6]
```diff
@@ -83,6 +84,52 @@ public void DeserializesObjectsWithIgnoreDefaults()
             Assert.IsNull(deserialized.color);
             Assert.IsNotNull(deserialized.canOffroad);
         }
+        
+        [TestMethod]
+        public void SerializesAndDeserializesWithCustomSerializers()
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:89]: Tests are cheap, you can write lots of them. This one test has 10 assertions. I'd rather see 10 tests, appropriately named with, (approximately) one assertion each.

File: Amazon.Ion.ObjectMapper.Test/TestObjects.cs [hunk 7]
```diff
@@ -286,4 +287,30 @@ public override string ToString()
             return "<Teacher>{ firstName: " + firstName + ", lastName: " + lastName + ", department: " + department + ", birthDate: " + birthDate + " }";
         }
     }
+    
+    public class UpperCaseStringIonSerializer : IonSerializer<string>
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper.Test/TestObjects.cs:291]: Does not belong in `TestObjects.cs`

File: Amazon.Ion.ObjectMapper/IonListSerializer.cs [hunk 8]
```diff
@@ -1,32 +1,57 @@
...
         private readonly IonSerializer serializer;
-        private readonly Type listType;
-        private readonly Type elementType;
-        private readonly bool isGenericList;
+        private Type listType;
+        private Type elementType;
+        private bool isGenericList;
 
-        public IonListSerializer(IonSerializer serializer, Type listType, Type elementType)
+        public IonListSerializer(IonSerializer serializer)
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonListSerializer.cs:15]: Why has this changed? How is this related to the custom serialization change?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 9]
```diff
@@ -11,7 +11,14 @@ public class IonObjectSerializer : IonSerializer<object>
         private const BindingFlags fieldBindings = BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Public;
         private readonly IonSerializer ionSerializer;
         private readonly IonSerializationOptions options;
-        private readonly Type targetType;
+        internal Type targetType;
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:14]: Why is this no visible and changeable? It should not be.

File: Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs [hunk 10]
```diff
@@ -190,11 +188,11 @@ public void Serialize(IIonWriter writer, string item)
     public class IonGuidSerializer : IonSerializer<Guid>
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs:188]: This change seems unrelated to the PR

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 11]
```diff
@@ -131,16 +135,54 @@ public interface IonSerializerFactory<T, TContext> where TContext : IonSerializa
...
+                {typeof(long), new IonLongSerializer()},
+                {typeof(float), new IonFloatSerializer()},
+                {typeof(double), new IonDoubleSerializer()},
+                {typeof(decimal), new IonDecimalSerializer()},
+                {typeof(BigDecimal), new IonBigDecimalSerializer()},
+                {typeof(SymbolToken), new IonSymbolSerializer()},
+                {typeof(DateTime), new IonDateTimeSerializer()},
+                {typeof(Guid), new IonGuidSerializer(this.options.AnnotateGuids)},
+            };
+            this.nullSerializer = new IonNullSerializer();
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:167]: I don't know why these are now fields.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 12]
```diff
@@ -131,16 +135,54 @@ public interface IonSerializerFactory<T, TContext> where TContext : IonSerializa
...
+            this.listSerializer = new IonListSerializer(this);
+            this.objectSerializer = new IonObjectSerializer(this);
+
+            if (this.options.IonSerializers != null)
+            {
+                foreach (var serializer in this.options.IonSerializers)
+                {
+                    if (ValidateCustomSerializer(serializer.Key, serializer.Value))
+                    {
+                        this.primitiveSerializers[serializer.Key] = serializer.Value;
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:178]: No I don't think this is right. Mutating the map of primitives is a bit icky. I would rather you just look in the map at the time you need to than do this replacement.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 13]
```diff
@@ -163,117 +205,34 @@ public void Serialize<T>(IIonWriter writer, T item)
...
             
-            if (listType.IsAssignableTo(typeof(System.Collections.IList)))
+            if (item is object)
             {
-                if (listType.IsGenericType)
-                {
-                    return new IonListSerializer(this, listType, listType.GetGenericArguments()[0]);
-                }
-                return new IonListSerializer(this, listType);
+                this.objectSerializer.targetType = type;
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:228]: Now this is no longer thread-safe.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 14]
```diff
@@ -288,87 +247,176 @@ public object Deserialize(IIonReader reader, Type type, IonType ionType)
...
-            throw new NotSupportedException("Don't know how to Deserialize this Ion data. Last IonType was: " + ionType);
+            throw new NotSupportedException($"Data with Ion type {ionType} is not supported for deserialization");
         }
 
         public T Deserialize<T>(IIonReader reader)
         {
             return (T) Deserialize(reader, typeof(T));
         }
+        
+        private bool ValidateCustomSerializer(Type type, dynamic serializer)
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:335]: Why is this validation necessary?

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 15]
```diff
@@ -288,87 +247,176 @@ public object Deserialize(IIonReader reader, Type type, IonType ionType)
...
+                return serializer is IonSerializer<Guid>;
+
+            return false;
+        }
+
+        private void SerializePrimitive(Type type, IIonWriter writer, object item)
+        {
+            var serializer = this.primitiveSerializers[type];
+            
+            if (type == typeof(bool))
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:385]: Please reinstate curly braces.

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 14 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:340]: Unfortunately we can't have a dictionary of `Type` and a generic of that type. So instead we are using a dictionary of `Type` with a `dynamic` object.
Since it's `dynamic`, we need to verify that the user provided a valid `IonSerializer<T>` for the given type.

File: Amazon.Ion.ObjectMapper.Test/TestObjects.cs (hunk 7 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/TestObjects.cs:325]: Why not? These serializers are just meant for testing the custom serializer functionality.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 3 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:224]: Ian seems to dislike switch statements.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs]: Added test ExceptionOnInvalidCustomSerializer

File: Amazon.Ion.ObjectMapper.Test/TestObjects.cs (hunk 4 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/TestObjects.cs:530]: Added tests for each primitive type custom serializer.

File: Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs (hunk 10 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs]: Undid change.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 11 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: No longer using these fields for these serializers. Instead we just spawn new ones each time we need one.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 12 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: Addressed.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 13 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: Undid change.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 15 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:403]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 9 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: The change was because the IonSerializer was changing the type. But that is no longer happening so this is back to `private readonly`.

File: Amazon.Ion.ObjectMapper/IonListSerializer.cs (hunk 8 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonListSerializer.cs]: This was done because we were allowing the IonSerializer to change the type but we are no longer doing that.

File: Amazon.Ion.ObjectMapper/IonListSerializer.cs [hunk 16]
```diff
@@ -1,35 +1,45 @@
...
-            this.isGenericList = true;
-        }
-
         public IonListSerializer(IonSerializer serializer, Type listType)
         {
             this.serializer = serializer;
             this.listType = listType;
-            this.isGenericList = false;
+
+            if (this.listType.IsArray)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonListSerializer.cs:20]: Is this change related to this PR? Is there a rebase that is needed?

File: Amazon.Ion.ObjectMapper/IonListSerializer.cs [hunk 17]
```diff
@@ -1,35 +1,45 @@
...
+                }
+            }
+            else
+            {
+                throw new NotSupportedException("Encountered an Ion list but the desired deserialized type was not an IList, it was: " + listType);
+            }
         }
 
-        public System.Collections.IList Deserialize(IIonReader reader)
+        public IList Deserialize(IIonReader reader)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonListSerializer.cs:39]: Can you revert these changes as they're unrelated to the PR and is not in code that you've changed, so it adds noise.

File: Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs [hunk 18]
```diff
@@ -65,7 +65,7 @@ public class IonLongSerializer : IonSerializer<long>
 
         public long Deserialize(IIonReader reader)
         {
-            return reader.IntValue();
+            return reader.LongValue();
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonPrimitiveSerializers.cs:68]: Related to this change? Seems like it should be a separate PR if it was discovered here.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 19]
```diff
@@ -127,6 +129,8 @@ public class IonSerializationOptions
         public string[] AnnotatedTypeAssemblies { get; init; } = new string[] {};
 
         public readonly bool PermissiveMode;
+
+        public Dictionary<Type, dynamic> IonSerializers { get; init; }
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:133]: Would this make more sense as a `Dictionary<Type, IonSerializer<dynamic>>` (if possible)?

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 14 above)
Comment (kylepbit) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs:329]: If you're able to change the dictionary to be `IonSerializer<dynamic>` instead of just `dynamic` I think think this check becomes trivial, since you'd only need to validate that the type of the serializer is already present for the primitives.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 20]
```diff
@@ -380,5 +325,150 @@ public T Deserialize<T>(IIonReader reader)
...
+            }
+
+            return false;
+        }
+
+        private void SerializePrimitive(Type type, IIonWriter writer, object item)
+        {
+            var serializer = this.GetPrimitiveSerializer(type);
+
+            if (type == typeof(bool))
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:403]: @alpian I'd assume this is hot code, wouldn't we prefer a switch statement for O(1) lookup (it reduces to a lookup table) instead of the fall-through if/else?

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 21]
```diff
@@ -137,15 +141,41 @@ public interface IonSerializerFactory<T, TContext> where TContext : IonSerializa
...
 
         public IonSerializer() : this(new IonSerializationOptions())
         {
-
         }
 
         public IonSerializer(IonSerializationOptions options)
         {
             this.options = options;
+            this.primitiveSerializers = new Dictionary<Type, dynamic>()
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:153]: If these are stateless, then we should just have a read-only static instance of the primitives rather than instantiating per-class.

**Comment by justing-bq** [comment]
Closing in favor of a new PR which will be posted soon.
